src/Datadog.Trace/AsyncLocalScopeManager.cs

```diff
@@ -10,51 +9,65 @@ internal class AsyncLocalScopeManager : IScopeManager
 
         private readonly AsyncLocalCompat<Scope> _activeScope = new AsyncLocalCompat<Scope>();
 
-        public event EventHandler<ScopeEventArgs> ScopeOpened;
+        public event EventHandler<SpanEventArgs> SpanOpened;
 
-        public event EventHandler<ScopeEventArgs> ScopeActivated;
+        public event EventHandler<SpanEventArgs> SpanActivated;
 
-        public event EventHandler<ScopeEventArgs> ScopeDeactivated;
+        public event EventHandler<SpanEventArgs> SpanDeactivated;
 
-        public event EventHandler<ScopeEventArgs> ScopeClosed;
+        public event EventHandler<SpanEventArgs> SpanClosed;
+
+        public event EventHandler<SpanEventArgs> TraceEnded;
 
         public Scope Active => _activeScope.Get();
 
         public Scope Activate(Span span, bool finishOnClose)
         {
-            var activeScope = _activeScope.Get();
-            var scope = new Scope(activeScope, span, this, finishOnClose);
-            var scopeOpenedArgs = new ScopeEventArgs(scope);
+            var newParent = Active;
+            var scope = new Scope(newParent, span, this, finishOnClose);
+            var scopeOpenedArgs = new SpanEventArgs(span);
+
+            SpanOpened?.Invoke(this, scopeOpenedArgs);
 
-            ScopeOpened?.Invoke(this, scopeOpenedArgs);
             _activeScope.Set(scope);
 
-            if (activeScope != null)
+            if (scope.Parent != null)
             {
-                ScopeDeactivated?.Invoke(this, new ScopeEventArgs(activeScope));
+                SpanDeactivated?.Invoke(this, new SpanEventArgs(newParent.Span));
             }
 
-            ScopeActivated?.Invoke(this, scopeOpenedArgs);
+            SpanActivated?.Invoke(this, scopeOpenedArgs);
+
             return scope;
         }
 
         public void Close(Scope scope)
         {
-            var current = _activeScope.Get();
+            var current = Active;
+            var isRootSpan = scope.Parent == null;
+
             if (current != null && current == scope)
             {
                 // if the scope that was just closed was the active scope,
                 // set its parent as the new active scope
                 _activeScope.Set(current.Parent);
-                ScopeDeactivated?.Invoke(this, new ScopeEventArgs(current));
+                SpanDeactivated?.Invoke(this, new SpanEventArgs(current.Span));
 
-                if (current.Parent != null)
+                if (!isRootSpan)
                 {
-                    ScopeActivated?.Invoke(this, new ScopeEventArgs(current.Parent));
+                    SpanActivated?.Invoke(this, new SpanEventArgs(current.Parent.Span));
                 }
             }
 
-            ScopeClosed?.Invoke(this, new ScopeEventArgs(scope));
+            if (current != null)
```

Comment:
I'm not sure that this condition is correct. The check for `current != null && current == scope` above is meant to make this `Close(Scope)` method a no-op if the argument is not the currently active scope. The code below can go into the `if` above.

At this point, if the `if` block is getting too long, we might want to reverse the condition and return, to reduce nesting. Something like:

```csharp
public void Close(Scope scope)
{
  var current = Active;

  if (current == null || current != scope)
  {
    // wrong scope, bail out
    return;
  }

  // ...do everything else...
}